STYLEGUIDE.md

```diff
@@ -0,0 +1,345 @@
+
+#### Why
+
+Documenting functions with [JSDoc](https://jsdoc.app/about-getting-started.html) helps PR reviewers better understand the intent of the function and also documents its use for other engineers using the code base. Better yet, [VS Code will automatically show in-line descriptions of a function](https://code.visualstudio.com/docs/languages/javascript#_jsdoc-support) when a user invokes that function in their editor.
+
+JSDocs should also clearly mention any major assumptions made about the function's input and output that may not be immediately obvious to an engineer unfamiliar with this part of the code base.
+
+#### Prefer
+
+```typescript
+/**
+ * Get the next applicable instance for upgrade. The function handles regions and cloud providers
+ * but does NOT work for Serverless clusters
+ * @param clusterDescription the backbone cluster description model
+ * @param providerOptionsModel the backbone provider options model
+ * @returns the next upgrade instance for this cluster
+ */
+export const getNextUpgradeInstance = (
+  clusterDescription: ClusterDescription,
+  providerOptionsModel: ProviderOptions,
+) => {
+  // ...
+};
+```
+
+#### Avoid
+
+```typescript
+export const getNextUpgradeInstance = (
+  clusterDescription: ClusterDescription,
+  providerOptionsModel: ProviderOptions,
+) => {
+  // ...
+};
+```
+
+---
+
+### Prefer creating the function in the highest scope, and passing in all required data.
+
+#### Why
+
+Keeps functions pure.
+
+---
+
+## <a id="Variable-naming">Variable Naming</a>
+
+### Acronyms are considered a single entity in camel-casing
+
+#### Why
+
+Some treat each letter in an acronym as its own entity when an acronyme is included in a var name. We'll instead consider the acronym as a single entity, and only capitalize the first letter of it (if that).
+
+#### Prefer
+
+```typescript
+let myWipValue = 'pending';
+```
+
+#### Avoid
+
+```typescript
+let myWIPValue = 'pending';
+```
+
+---
+
+### Use UPPERCASE for static constants, and refactor to a shared file if needed.
+
+---
+
+### Use camelCase for most javascript functions and variables
+
+---
+
+### Use “be” verbs for boolean types. (e.g. should`/`can`/`does`/`has`)
+
+---
+
+## <a id="Variable-placement">Variable placement</a>
+
+### Avoid inline declaration of static variables when they can be moved to top of file or refactored into new file
+
+---
+
+### Prefer placing let/const variables at the beginning of their scope.
+
+---
+
+### Prefer const to let whenever possible.
+
+---
+
+## <a id="Whitespace-and-line-breaks">Whitespace and line breaks</a>
+
+### Use a space before every prop defined in an interface
+
+---
+
+## <a id="JSX-React">JSX/React</a>
+
+### Prefer factoring out any more-than-1-line event handlers into their own functions
+
+#### Why
+
+Minimizes inline javascript logic within
+
+#### Prefer
+
+```typescript
+const handleClick = (e) => {
+    e.preventDefault()
+    setState(curr => !curr)
+}
+
+<button onClick={handleClick}>
```

Comment:
Might also be nice to mention matching the event handler names using `on` to `handle`

Reply:
Good feedback, added!